ReSharper.FSharp/src/FSharp.Psi.Features/src/Intentions/FunctionAnnotationAction.fs

```diff
@@ -0,0 +1,133 @@
+open JetBrains.ReSharper.Psi.ExtensionsAPI
+open JetBrains.ReSharper.Psi.ExtensionsAPI.Tree
+open JetBrains.ReSharper.Psi.Tree
+open JetBrains.ReSharper.Resources.Shell
+
+type private ParameterToModify = {
+    ParameterNameNode : ILocalReferencePat
+    ParameterNodeInSignature : ISynPat
+    FSharpParameters : FSharpParameter list
+}
+
+type private ParameterNodeWithType = {
+    Node : ITreeNode
+    FSharpParameters : FSharpParameter list
+}
+
+[<ContextAction(Name = "AnnotateFunction", Group = "F#", Description = "Annotate function with parameter types and return type")>]
+type FunctionAnnotationAction(dataProvider: FSharpContextActionDataProvider) =
+    inherit ContextActionBase()
+
+    let rec getTypeDisplayName (fsType : FSharpType) : string =
+        let typeDef = if fsType.HasTypeDefinition then fsType.TypeDefinition.DisplayName else ""
+        let genericParam = if fsType.IsGenericParameter then "'" + fsType.GenericParameter.DisplayName else ""
+        let genericArgs = if fsType.HasTypeDefinition then fsType.GenericArguments |> Seq.map getTypeDisplayName |> Seq.toList else List.empty
+        let tupleArgs = if fsType.IsTupleType then fsType.GenericArguments |> Seq.map getTypeDisplayName |> String.concat "*" else ""
+        
+        [genericParam; typeDef; tupleArgs]
+        |> List.append genericArgs
+        |> List.filter (fun x -> String.length x <> 0)
+        |> String.concat " "
+    
+    override x.Text = "Annotate function with parameter types and return type"
+    
+    override x.IsAvailable _ =
+        let letBindings = dataProvider.GetSelectedElement<ILetBindings>()
+        if isNull letBindings || letBindings.IsRecursive then false else
+
+        let letToken = letBindings.LetOrUseToken
+        if isNull letToken then false else
+
+        let bindings = letBindings.Bindings
+        if bindings.Count <> 1 then false else
+
+        match bindings.[0].HeadPattern.As<INamedPat>() with
+        | null -> false
+        | namedPat ->
+
+        match namedPat.Identifier with
+        | null -> false
+        | _identifier -> true
+        
+    override x.ExecutePsiTransaction(_, _) =
+        let letExpr = dataProvider.GetSelectedElement<ILetBindings>()
+        use _writeCookie = WriteLockCookie.Create(letExpr.IsPhysical())
+        use _disableFormatter = new DisableCodeFormatter()
+        let binding = letExpr.Bindings |> Seq.exactlyOne
+        match binding.HeadPattern.As<INamedPat>() with
+        | null -> null
+        | namedPat ->
+            
+        let methodSymbol = namedPat.GetFSharpSymbol()
+        let fSharpFunction =
+            match methodSymbol with
+            | :? FSharpMemberOrFunctionOrValue as x -> x
+            | _ -> failwith "Expected function here"
+        let treeParameters =
+            namedPat.Children()
+            |> Seq.choose(
+                function
+                 | :? ILocalReferencePat as ref -> ref :> ITreeNode |> Some
```

Comment:
I'd suggest using generated getters for particular pattern types, something like:
```fsharp
match namedPat with
| :? IParameterOwnerPat as paramOwner -> paramOwner.Parameters
| :? IReferencePat -> namedPat
...
```